File: 54remitflow-unified-platform/backend/python-services/shared/resilience.py
"""
Resilience utilities for 54Agent Banking Platform

Provides circuit breakers, retry with exponential backoff, and explicit
timeouts for all outbound HTTP calls.

Usage::

    from shared.resilience import resilient_client, circuit_breaker

    async with resilient_client() as client:
        resp = await client.get("https://api.example.com/health")

    @circuit_breaker("payment-gateway")
    async def call_gateway(payload):
        ...
"""

import os
import time
import asyncio
import logging
import random
from typing import Optional, Dict, Any, Callable, TypeVar, Awaitable
from functools import wraps
from enum import Enum

import httpx

logger = logging.getLogger("platform.resilience")

T = TypeVar("T")

DEFAULT_CONNECT_TIMEOUT = float(os.getenv("HTTP_CONNECT_TIMEOUT", "5"))
DEFAULT_READ_TIMEOUT = float(os.getenv("HTTP_READ_TIMEOUT", "30"))
DEFAULT_RETRIES = int(os.getenv("HTTP_RETRIES", "3"))
DEFAULT_BACKOFF_BASE = float(os.getenv("HTTP_BACKOFF_BASE", "0.5"))
DEFAULT_BACKOFF_MAX = float(os.getenv("HTTP_BACKOFF_MAX", "30"))


def resilient_client(
    *,
    connect_timeout: float = DEFAULT_CONNECT_TIMEOUT,
    read_timeout: float = DEFAULT_READ_TIMEOUT,
    retries: int = DEFAULT_RETRIES,
    headers: Optional[Dict[str, str]] = None,
) -> httpx.AsyncClient:
    transport = httpx.AsyncHTTPTransport(retries=retries)
    timeout = httpx.Timeout(connect_timeout, read=read_timeout)
    return httpx.AsyncClient(
        transport=transport,
        timeout=timeout,
        headers=headers or {},
    )


async def retry_async(
    fn: Callable[..., Awaitable[T]],
    *args: Any,
    retries: int = DEFAULT_RETRIES,
    backoff_base: float = DEFAULT_BACKOFF_BASE,
    backoff_max: float = DEFAULT_BACKOFF_MAX,
    retryable_exceptions: tuple = (httpx.HTTPStatusError, httpx.ConnectError, httpx.TimeoutException, ConnectionError, OSError),
    **kwargs: Any,
) -> T:
    last_exc: Optional[BaseException] = None
    for attempt in range(1, retries + 1):
        try:
            return await fn(*args, **kwargs)
        except retryable_exceptions as exc:
            last_exc = exc
            if attempt == retries:
                break
            delay = min(backoff_base * (2 ** (attempt - 1)), backoff_max)
            jitter = random.uniform(0, delay * 0.25)
            logger.warning(
                "Retry %d/%d for %s after %.2fs: %s",
                attempt, retries, fn.__name__, delay + jitter, exc,
            )
            await asyncio.sleep(delay + jitter)
    raise last_exc  # type: ignore[misc]


class CircuitState(Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


class _CircuitBreaker:
    __slots__ = (
        "name", "failure_threshold", "recovery_timeout", "half_open_max",
        "_state", "_failure_count", "_success_count", "_last_failure_time",
    )

    def __init__(
        self,
        name: str,
        failure_threshold: int = 5,
        recovery_timeout: float = 30.0,
        half_open_max: int = 1,
    ):
        self.name = name
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.half_open_max = half_open_max
        self._state = CircuitState.CLOSED
        self._failure_count = 0
        self._success_count = 0
        self._last_failure_time = 0.0

    @property
    def state(self) -> CircuitState:
        if self._state == CircuitState.OPEN:
            if time.monotonic() - self._last_failure_time >= self.recovery_timeout:
                self._state = CircuitState.HALF_OPEN
                self._success_count = 0
        return self._state

    def record_success(self) -> None:
        if self._state == CircuitState.HALF_OPEN:
            self._success_count += 1
            if self._success_count >= self.half_open_max:
                self._state = CircuitState.CLOSED
                self._failure_count = 0
                logger.info("Circuit %s CLOSED (recovered)", self.name)
        else:
            self._failure_count = 0

    def record_failure(self) -> None:
        self._failure_count += 1
        self._last_failure_time = time.monotonic()
        if self._failure_count >= self.failure_threshold:
            self._state = CircuitState.OPEN
            logger.warning("Circuit %s OPEN after %d failures", self.name, self._failure_count)

    def allow_request(self) -> bool:
        s = self.state
        if s == CircuitState.CLOSED:
            return True
        if s == CircuitState.HALF_OPEN:
            return True
        return False


_breakers: Dict[str, _CircuitBreaker] = {}


def get_breaker(
    name: str,
    failure_threshold: int = 5,
    recovery_timeout: float = 30.0,
) -> _CircuitBreaker:
    if name not in _breakers:
        _breakers[name] = _CircuitBreaker(
            name=name,
            failure_threshold=failure_threshold,
            recovery_timeout=recovery_timeout,
        )
    return _breakers[name]


def circuit_breaker(
    name: str,
    failure_threshold: int = 5,
    recovery_timeout: float = 30.0,
):
    def decorator(fn: Callable[..., Awaitable[T]]) -> Callable[..., Awaitable[T]]:
        cb = get_breaker(name, failure_threshold, recovery_timeout)

        @wraps(fn)
        async def wrapper(*args: Any, **kwargs: Any) -> T:
            if not cb.allow_request():
                raise HTTPCircuitOpenError(f"Circuit '{name}' is OPEN")
            try:
                result = await fn(*args, **kwargs)
                cb.record_success()
                return result
            except Exception as exc:
                cb.record_failure()
                raise
        return wrapper
    return decorator


class HTTPCircuitOpenError(Exception):
    pass

File: 54remitflow-unified-platform/backend/go-services/shared/resilience/resilience.go
package resilience

import (
	"context"
	"fmt"
	"math"
	"math/rand"
	"net/http"
	"os"
	"strconv"
	"sync"
	"time"
)

func getEnvInt(key string, fallback int) int {
	if v := os.Getenv(key); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			return n
		}
	}
	return fallback
}

func getEnvFloat(key string, fallback float64) float64 {
	if v := os.Getenv(key); v != "" {
		if f, err := strconv.ParseFloat(v, 64); err == nil {
			return f
		}
	}
	return fallback
}

type CircuitState int

const (
	Closed CircuitState = iota
	Open
	HalfOpen
)

type CircuitBreaker struct {
	mu               sync.Mutex
	state            CircuitState
	failures         int
	successes        int
	failureThreshold int
	successThreshold int
	timeout          time.Duration
	lastFailure      time.Time
}

func NewCircuitBreaker() *CircuitBreaker {
	return &CircuitBreaker{
		state:            Closed,
		failureThreshold: getEnvInt("CB_FAILURE_THRESHOLD", 5),
		successThreshold: getEnvInt("CB_SUCCESS_THRESHOLD", 2),
		timeout:          time.Duration(getEnvInt("CB_TIMEOUT_SECONDS", 30)) * time.Second,
	}
}

func (cb *CircuitBreaker) Allow() bool {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	switch cb.state {
	case Closed:
		return true
	case Open:
		if time.Since(cb.lastFailure) > cb.timeout {
			cb.state = HalfOpen
			cb.successes = 0
			return true
		}
		return false
	case HalfOpen:
		return true
	}
	return false
}

func (cb *CircuitBreaker) RecordSuccess() {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.failures = 0
	if cb.state == HalfOpen {
		cb.successes++
		if cb.successes >= cb.successThreshold {
			cb.state = Closed
		}
	}
}

func (cb *CircuitBreaker) RecordFailure() {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.failures++
	cb.lastFailure = time.Now()
	if cb.failures >= cb.failureThreshold {
		cb.state = Open
	}
}

func (cb *CircuitBreaker) State() CircuitState {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	return cb.state
}

type RetryConfig struct {
	MaxRetries  int
	BackoffBase float64
	BackoffMax  float64
}

func DefaultRetryConfig() RetryConfig {
	return RetryConfig{
		MaxRetries:  getEnvInt("HTTP_RETRIES", 3),
		BackoffBase: getEnvFloat("HTTP_BACKOFF_BASE", 0.5),
		BackoffMax:  getEnvFloat("HTTP_BACKOFF_MAX", 30.0),
	}
}

func RetryDo(ctx context.Context, cfg RetryConfig, fn func() error) error {
	var lastErr error
	for attempt := 0; attempt <= cfg.MaxRetries; attempt++ {
		lastErr = fn()
		if lastErr == nil {
			return nil
		}
		if attempt < cfg.MaxRetries {
			delay := math.Min(cfg.BackoffBase*math.Pow(2, float64(attempt)), cfg.BackoffMax)
			jitter := delay * 0.1 * rand.Float64()
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(time.Duration((delay + jitter) * float64(time.Second))):
			}
		}
	}
	return fmt.Errorf("exhausted %d retries: %w", cfg.MaxRetries, lastErr)
}

func ResilientClient() *http.Client {
	connectTimeout := time.Duration(getEnvInt("HTTP_CONNECT_TIMEOUT", 5)) * time.Second
	readTimeout := time.Duration(getEnvInt("HTTP_READ_TIMEOUT", 30)) * time.Second
	return &http.Client{
		Timeout: connectTimeout + readTimeout,
		Transport: &http.Transport{
			MaxIdleConns:        100,
			MaxIdleConnsPerHost: 10,
			IdleConnTimeout:     90 * time.Second,
		},
	}
}
